src/sycltest/SYCLCore/chooseDevice.cc

```diff
@@ -0,0 +1,32 @@
+#include <vector>
+
+#include <CL/sycl.hpp>
+
+#include "chooseDevice.h"
+
+namespace cms::sycltools {
+  std::vector<sycl::device> const& enumerateDevices(bool verbose) {
+    static const std::vector<sycl::device> devices = sycl::device::get_devices(sycl::info::device_type::all);
+    if (verbose) {
+      std::cerr << "Found " << devices.size() << " SYCL devices:" << std::endl;
+      for (auto const& device: devices)
+        std::cerr << "  - " << device.get_info<cl::sycl::info::device::name>() << std::endl;
+      std::cerr << std::endl;
+    }
+    return devices;
+  }
+
+  sycl::device chooseDevice(edm::StreamID id) {
+    auto const& devices = enumerateDevices();
+
+    // For startes we "statically" assign the device based on
+    // edm::Stream number. This is suboptimal if the number of
+    // edm::Streams is not a multiple of the number of CUDA devices
+    // (and even then there is no load balancing).
+    //
+    // TODO: improve the "assignment" logic
+    auto const& device = devices[id % devices.size()];
```

Comment:
Do I understand correctly that on a machine with CPU and GPU, this will alternate between them?

Reply:
Yes, currently this will cycle through all available devices: "host device", OpenCL cpu driver, OpenCL gpu driver, maybe also Level 0 gpu driver (i.e. the gpu is overcommited), FPGA emulator device...

The device discovery and selection clearly needs more work.